### PR TITLE
Add a `testMode` Map option to allow for usage in tests without an access token

### PR DESCRIPTION
**Pull request by arindam1993**
This PR adds a new `Map` constructor option called `testMode`. When enabled it allows for the usage of the map in automated tests without an accessToken with custom self-hosted test fixtures.

In this mode, the Map produces no rendered output on the canvas, but still retains all state for features and the camera. This means mocked mouse and UI events will work as expected.

In order to "test" the `testMode` I migrated our unit and selenium tests to use it, and removed the custom access tokens that were hardcoded in there.

## Launch Checklist

 - [x] briefly describe the changes in this PR
 - [x] write tests for all new functionality
 - [x] document any changes to public APIs
 - [x] manually test the debug page
 - [x] apply changelog label ('bug', 'feature', 'docs', etc) or use the label 'skip changelog'
 - [x] add an entry inside this element for inclusion in the `mapbox-gl-js` changelog: `<changelog></changelog>`

### CODE REVIEW

File: src/render/draw_sky.js [hunk 1]
```diff
@@ -95,7 +94,8 @@ function drawSkyboxFromCapture(painter: Painter, layer: SkyLayer, depthMode: Dep
         layer.skyboxGeometry.indexBuffer, layer.skyboxGeometry.segment);
 }
 
-function drawSkyboxFace(context: Context, layer: SkyLayer, program: Program<*>, faceRotate: mat4, sunDirection: vec3, i: number) {
+function drawSkyboxFace(painter: Painter, layer: SkyLayer, program: Program<*>, faceRotate: mat4, sunDirection: vec3, i: number) {
```
Comment (karimnaaji) [inline src/render/draw_sky.js:97]: Is it necessary to extend this to painter or is it a leftover?

File: src/ui/map.js [hunk 2]
```diff
@@ -2664,13 +2671,16 @@ class Map extends Camera {
             if (err) {
                 // throwing an error here will cause the callback to be called again unnecessarily
                 if (err.message === AUTH_ERR_MSG || err.status === 401) {
-                    console.error('Error: A valid Mapbox access token is required to use Mapbox GL JS. To create an account or a new access token, visit https://account.mapbox.com/');
-                    browser.setErrorState();
                     const gl = this.painter.context.gl;
+                    storeAuthState(gl, false);
```
Comment (karimnaaji) [inline src/ui/map.js:2675]: One question about storeAuthState/removeAuthState, how do we ensure storeAuthState always followed by removeAuthState? If this is even possible in practice, what would be the behavior in that case?

File: src/util/mapbox.js [hunk 3]
```diff
@@ -403,7 +407,7 @@ export class MapLoadEvent extends TelemetryEvent {
             if (customAccessToken || config.ACCESS_TOKEN) {
                 this.queueRequest({id: mapId, timestamp: Date.now()}, customAccessToken);
             } else {
-                this.errorCb(new Error('A valid Mapbox access token is required to use Mapbox GL JS. To create an account or a new access token, visit https://account.mapbox.com/'));
+                this.errorCb(new Error(AUTH_ERR_MSG));
```
Comment (karimnaaji) [inline src/util/mapbox.js:410]: Is the message the same as what is stored in AUTH_ERR_MSG?

File: src/ui/map.js [hunk 4]
```diff
@@ -2664,13 +2671,16 @@ class Map extends Camera {
             if (err) {
                 // throwing an error here will cause the callback to be called again unnecessarily
                 if (err.message === AUTH_ERR_MSG || err.status === 401) {
-                    console.error('Error: A valid Mapbox access token is required to use Mapbox GL JS. To create an account or a new access token, visit https://account.mapbox.com/');
```
[the remark this answers is not in the document]
Comment (arindam1993) [reply src/ui/map.js:2667]: Yep, we handle it in `Evented` https://github.com/mapbox/mapbox-gl-js/blob/89fa6737e54cba7b93fb87123399817631502c6c/src/util/evented.js#L145

An extra console.error log was unnecessary and was duplicating the error logged by Eveneted.

File: src/ui/map.js (hunk 2 above)
Comment (arindam1993) [reply src/ui/map.js:2675]: `removeAuthState` has to be called only when the map is removed. I don't think we can gurantee it other than ensuring we never deleting the line in `Map#remove`

Comment (karimnaaji) [reply src/ui/map.js:2675]: Should we go over all stored states and assert if anything is leftover?

File: src/util/mapbox.js (hunk 3 above)
Comment (arindam1993) [reply src/util/mapbox.js:410]: No it isnt, `AUTH_ERR_MSG` is just a code that says "NO_ACCESS_TOKEN". This error is never thrown/logged , its used by the caller(`Map#_authenticate`) to un-authenticate the map in specific cases only.

File: src/render/program.js [hunk 5]
```diff
@@ -162,6 +164,8 @@ class Program<Us: UniformBindings> {
         const gl = context.gl;
 
         if (this.failedToCreate) return;
+        // Following line is billing related code. Do not change. See LICENSE.txt
+        if (!isMapAuthenticated(gl)) return;
```
Comment (karimnaaji) [inline src/render/program.js:168]: Could this be done at a higher level than in the program? I remember you checked the overhead of this call, is it fairly negligible?

File: src/ui/map.js [hunk 6]
```diff
@@ -2626,6 +2632,7 @@ class Map extends Camera {
             if (!this.isMoving() && this.loaded()) {
                 this.fire(new Event('idle'));
                 if (this._isInitialLoad) {
+                    // Following line is billing related code. Do not change. See LICENSE.txt
```
Comment (rreusser) [inline src/ui/map.js:2635]: Maybe lift this statement outside the conditional? `s/line/block/`? No strong feeling here, as long as the intent is clear.

**Review by rreusser — APPROVED** [review]

File: src/ui/map.js (hunk 6 above)
Comment (rreusser) [reply src/ui/map.js:2635]: Edit: Now that I re-read, the intent is clear. Kindly disregard.

**Review by karimnaaji — APPROVED** [review]
LGTM, thanks for moving the auth check up to Painter.

File: src/ui/map.js [hunk 7]
```diff
@@ -104,6 +104,7 @@ type MapOptions = {
     maxTileCacheSize?: number,
     transformRequest?: RequestTransformFunction,
     accessToken: string,
+    testMode: ?boolean,
```
Comment (TrySound) [inline src/ui/map.js:107]: Shouldn't it be optional, not just nullable?

**Pull request merged**